Added service for mastery data

LennardF1989 opened the pull request:
~~Depends on: #25~~ Added by RDIL, but this is not true. It has portions of which Evergreen can benefit, however.

This PR includes:
- A service to handle/retrieve mastery data
- Mastery data for all locations, including Evergreen (SNUG).
- Added initial support for profile progression on the Player Profile screen as well as when viewing mastery of locations.

This PR does not include:
- Actual progression saving/drops on /MissionEnd, even though I already sorted out mastery drops as part of my custom scoring plugin for Evergreen and the storing of that progression per location in the player profile. I will backport that to general purpose later.

General note:
- This has been tested on Hitman 3 for Epic, and technically the Steam version though the playtest. 
- Most likely will work on Hitman 1 and Hitman 2. Otherwise it would be rather easy to exclude this based on game version.

RDIL commented:
That was unexpected.

AnthonyFuller commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,124 @@
+        locationParentId: string,
+        gameVersion: GameVersion,
+        userId: string,
+    ): unknown[] {
```
These should probably be properly typed.

AnthonyFuller commented on components/controller.ts:
```diff
@@ -831,6 +840,26 @@ export class Controller {
         }
     }
 
+    private async _loadMasteryData(): Promise<void> {
```
This should probably be combined with `_loadChallenges` into a `_loadResources` function.

LennardF1989 replied:
See comment above. I'm all for going for a generic resource package format. If this were to become loadResources that makes sense. However, from a Single Responsibility point of view while it does load resources, it also does the registration for the specific type of the data it found. I would rather go for a _loadResources-function that does the byte reading, and then two functions (as callbacks?) to handle the actual processing.

RDIL commented on contractdata/CARPATHIAN/_CARPATHIAN_MASTERY.json:
```diff
@@ -0,0 +1,192 @@
+            {
+                "IsLevelMarker": false,
+                "Unlockable": {
+                    "Properties": {
+                        "Location": "LOCATION_PARENT_TRAPPED",
+                        "RepositoryId": "f21b1a1a-bec9-4a06-9a4a-d615e9721e3a",
+                        "UnlockOrder": 1,
+                        "LoadoutSettings": {
+                            "GearSlotsEnabledCount": 0,
+                            "GearSlotsAllowContainers": false,
+                            "ConcealedWeaponSlotEnabled": false
+                        },
+                        "Equip": [
+                            "TOKEN_OUTFIT_TRAPPED_WOLVERINE_SUIT"
+                        ]
+                    },
+                    "Rarity": null,
+                    "DisplayNameLocKey": "UI_STARTING_LOCATION_TRAPPED_WOLVERINE_TRAIN_LAB_NAME",
+                    "Id": "STARTING_LOCATION_TRAPPED_WOLVERINE_TRAIN_LAB",
+                    "Type": "access",
+                    "Subtype": "startinglocation",
+                    "GameAsset": null,
+                    "ImageId": "",
+                    "RMTPrice": -1,
+                    "GamePrice": -1,
+                    "IsPurchasable": false,
+                    "IsPublished": true,
+                    "IsDroppable": false,
+                    "Capabilities": [],
+                    "Qualities": {},
```
These are duplicated from `allunlockables` - we don't want to waste this much memory on duplicated data.

LennardF1989 replied:
Agree, but see my comment under Anthony's - I think we have to pick a lesser evil here. I'm all for de-deplication though.

RDIL commented on packaging/buildTasks.mjs:
```diff
@@ -34,32 +34,38 @@ async function readJson(filePath) {
+    const targetName = createHash("md5").update(name).digest("hex")
+
+    await writeFile(
+        join(resources, `${targetName}.crp`),
```
CRP = challenge resource package, this is a new format, and it doesn't really make sense to have it as the same extension if it stores different data.

LennardF1989 replied:
See my comment on Anthony's response to that.

AnthonyFuller reviewed (changes requested):
All in all a good PR and an especially welcome one! Thanks a lot!

RDIL commented on static/EvergreenMasteryDataForLocationTemplate.json:
```diff
@@ -0,0 +1,688 @@
+{
+    "template": {
```
For the storage of templates, we want the template to be top-level.

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,124 @@
+            return evergreenMasteryDataTemplate
+        }
+
+        return {
```
Building on this - if it is, the return type doesn't make sense to be called `EvergreenMasteryDataTemplate`.

LennardF1989 replied:
It's currently only possible to see mastery data when in the Safehouse ingame menu. Other location do not do this (yet?). For Evergreen, it has to stay, so I would propose a better name for this then.

AnthonyFuller replied:
I checked this in Retail files, the `MasteryDataForLocation` endpoint is referenced elsewhere irregardless of EG.

RDIL replied:
Fairly sure you can in Sniper Assassin too.

AnthonyFuller replied:
It's specifically used when pressed tiles displayed in Career -> Mastery on the main menu, good reuse.

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,124 @@
+import { getConfig } from "components/configSwizzleManager"
```
Consistency - should be `../components/`

AnthonyFuller commented on packaging/buildTasks.mjs:
```diff
@@ -78,9 +84,15 @@ export async function packContractsAndChallenges() {
         const json = await readJson(path)
 
-        if (filename.startsWith("_")) {
+        if (filename.endsWith("_CHALLENGES.json")) {
```
This may be invalid as it is probably required to detect if the extension should be `crp` or `mrp`.

LennardF1989 replied:
I thought about adding extension support instead, but the easier solution was going with the existing packing logic. 

Two choices:
- Implement multiple extension support, de-duplicate code. Everything could go into a single folder.
- Go for a more generic extension, keep multiple folder logic.

AnthonyFuller replied:
The latter is probably the better of the two options.

LennardF1989 commented on contractdata/AMBROSE/_AMBROSE_MASTERY.json:
```diff
@@ -0,0 +1,598 @@
+[
```
The completion data is reset and includes the MaxLevel - this is not always 20, but sometimes 5 and in case of evergreen 25 and in the full release even more.

The reason I have the format as is, is because it significantly simplifies the loading of the data into the proper locations. Otherwise you will either end up:
- Generating this exact output when loading the .csp file with the data (startup overhead, no memory difference)
- Generating this data every time you need it (aka: mapping - no startup overhead, less static memory)

So it's a give or take situation. From a maintainability perspective a smaller revised format is easier, but it comes at a cost.

AnthonyFuller replied:
Adding onto this, `TypeLocaKey` can be generated from the unlockable, so here's an updated format:
```json
{
    "Id": "LOCATION_PARENT_ROCKY",
    "Levels": [
        {
            "Level": 2,
            "Unlockable": "AGENCYPICKUP_ROCKY_SMALL_BIGCENTRALTREE",
            "IsLevelMarker": false
        }
    ]
}
```

AnthonyFuller replied:
> The completion data is reset and includes the MaxLevel
What do you measn by "reset", and "MaxLevel" can be added by a property, just like Id.

I'm sure @RDIL can comment about the generation part, I think having the smaller format is more beneficial as we're not storing the _exact_ same data twice, and it's actually better for maintaining it as if the unlockable changes between updates (i.e. a suit or weapon), we'd have to update the mastery json too while we'd normally just update the one `allunlockables`.

RDIL replied:
Unless the duplication has a very good reason for being present, I'd like to avoid it.  Very likely, reducing memory by, say, 10 megabytes, is a better improvement then saving less then say 50 milliseconds.

AnthonyFuller replied:
It also isn't something that would be ran often so it wouldn't impact much.

LennardF1989 replied:
> What do you mean by "reset", and "MaxLevel" can be added by a property, just like Id.
The MaxLevel determines the maximum level of the mastery without having the scan through the mastery data to grab the highest available level.

But I understand where you guys want to go. Runtime generation of the full data on-demand it is.

LennardF1989 commented on packaging/buildTasks.mjs:
```diff
@@ -34,32 +34,38 @@ async function readJson(filePath) {
+    )
+}
+
+export async function packContractsAndChallenges() {
```
Agree, but see above comment. I guess it comes down on how we want to handle packing this data (and potentially more in the future).

LennardF1989 commented:
Added new commits with the requested changes. Purposely didn't squash so you can easily spot the changes.

While I've fixed a performance issue now that everything is generated runtime, I must add it happens quite often: Every time you go into a Destination. So when you are casually browsing to decide the contract to play, you are constantly doing a lot of calculations to build this object with drops. 

If that is acceptable in order to safe some memory, then I'm okay with it.

LennardF1989 commented on components/index.ts:
```diff
@@ -505,6 +505,7 @@ function startServer(options: { hmr: boolean; pluginDevHost: boolean }): void {
         "contractSessions",
         "userdata",
         "contracts",
+        "plugins",
```
This is actually a bug-fix for the hot-reload. I will remove those when I squash after the second round of checks.

AnthonyFuller commented on contractdata/AMBROSE/_AMBROSE_MASTERY.json:
```diff
@@ -0,0 +1,85 @@
+{
+  "Id": "LOCATION_PARENT_ROCKY",
```
General comment about the mastery data:
If you require max level, you can add it under here e.g.
```json
{
    "Id": "LOCATION_PARENT_ROCKY",
    "MaxLevel": 20,
    "Drops": [...]
}
```

LennardF1989 replied:
I determined the MaxLevel based on the highest mastery drop, so not needed. They all match up. Only exception would be Evergreen, where there is a Mastery up til 25, but we could technically make it go indefinately (especially since the Safehouse has more unlocks in it post level-25).

AnthonyFuller replied:
Adding MaxLevel as a property would reduce the time seeking for the max level, having it default to 20 and an optional property sounds good. Only exceptions are Evergreen and Carpathian.

AnthonyFuller commented on components/contracts/dataGen.ts:
```diff
@@ -97,24 +97,34 @@ export function generateCompletionData(
+    if (!completionData) {
+        return <CompletionData>{
+            Level: 20,
+            MaxLevel: 20,
```
See max level comment in the mastery data. Also, if this is default completion data, why is it max level, and if it's meant to be fully completed, why is XpLeft not 0?

LennardF1989 replied:
This is literally what it was originally (just check the v5 branch). But you are correct, it doesn't make sense.

AnthonyFuller replied:
The reason v5 does this is because it doesn't have a progression system so you can't get mastery, it used to be Level 1, but it was changed to allow easy comparison with official server responses when developing the challenge system.

AnthonyFuller commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,196 @@
+            gameVersion,
+        )
+
+        const evergreenMasteryDataTemplate: unknown =
```
Rename this variable as it's not EG specific.

AnthonyFuller reviewed (changes requested):
This is definitely getting closer to how I envisioned a mastery system. Just a few final things.

LennardF1989 commented:
Since all changes are most likely fully conform agreements, I went ahead and squashed the branch already for a final review.

AnthonyFuller commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,198 @@
+
+        const maxLevel =
+            masteryData.MaxLevel ||
+            masteryData.Drops.reduce((currentMax, drop) => {
```
Do you think having this just default to 20 would be more beneficial? As there are only 2 exceptions (Evergreen and Carpathian being 25-100 and 5 respectively).

LennardF1989 replied:
Yes, I would go with a default of 20 and the let the JSON tell you if it's supposed to be something else.

AnthonyFuller reviewed (changes requested):
We're almost there, I can feel it! Just one final thing.

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,198 @@
+        //Gather all required data
+        const lowerCaseLocationParentId = locationParentId.toLowerCase()
+
+        if (
+            !userProfile.Extensions.progression.Locations[
+                lowerCaseLocationParentId
+            ]
+        ) {
+            userProfile.Extensions.progression.Locations[
+                lowerCaseLocationParentId
+            ] = {
+                Xp: 0,
+                Level: 1,
+            }
+        }
```
```suggestion
        userProfile.Extensions.progression.Locations[
                lowerCaseLocationParentId
        ] ??= {
            Xp: 0,
            Level: 1,
        }
```

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,198 @@
+            userId,
+        )
+
+        return <MasteryDataTemplate>{
```
Why is this cast needed?

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,198 @@
+            }
+        })
+
+        return <MasteryData[]>[
```
Same here.

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,198 @@
+        }).map((drop) => {
+            const unlockable: Unlockable = unlockableMap.get(drop.Id)
+
+            return <MasteryDrop>{
```
Same here.

RDIL commented on components/contracts/dataGen.ts:
```diff
@@ -97,24 +97,39 @@ export function generateCompletionData(
+            "Could not get CompletionData for location ${locationId}",
+        )
+
+        return <CompletionData>{
```
Same here.

AnthonyFuller replied:
This cast is still here!

RDIL commented on components/candle/masteryService.ts:
```diff
@@ -0,0 +1,198 @@
+import { xpRequiredForLevel } from "../../components/utils"
+
+export class MasteryService {
+    // @ts-expect-error Unused
```
Why have this if unused?

AnthonyFuller commented on contractdata/SNUG/_SNUG_MASTERY.json:
```diff
@@ -0,0 +1,186 @@
+{
```
Remove this file, it will be added by the Evergreen branch/PR.

AnthonyFuller commented on components/contracts/dataGen.ts:
```diff
@@ -216,12 +231,14 @@ export function generateUserCentric(
     objectives: MissionManifestObjective[],
     gameChangers: string[],
     displayOrder: GroupObjectiveDisplayOrderItem[],
+    IsEvergreenSafehouse = false,
```
How did this end up here?

LennardF1989 replied:
Oops, due to the switch-a-roo after moving it away from the evergreen branch :)

AnthonyFuller reviewed (changes requested):
A few final final things!

AnthonyFuller reviewed (approved):
LGTM. Thank you so much for implementing this!
